### ISSUE TITLE
Document serving static files with nginx (with SSR backend)

### ISSUE THREAD

**Issue by ishitatsuyuki** [nuxt/nuxt.js#2517]
Currently Nuxt.js uses `serve-static` as a middleware for static files. This can be offloaded to nginx for efficiency.

Basically, the code [here](https://github.com/nuxt/nuxt.js/blob/9fc2bace443a267bdf709928a9d6096f4fbc67c8/lib/core/renderer.js#L234-L248) needs to be documented.

<div align="right"><sub><em>This feature request is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c2186">#c2186</a>)</em></sub></div>

**Comment by ishitatsuyuki** [comment]
These too:
https://github.com/nuxt/nuxt.js/blob/9fc2bace443a267bdf709928a9d6096f4fbc67c8/lib/core/renderer.js#L388-L412

**Comment by clarkdo** [comment]
Please refer https://nuxtjs.org/api/configuration-render#static

**Comment by ishitatsuyuki** [comment]
Wrong, please reopen. I'm not asking for configuring `serve-static`, I want to serve the static files on nginx.

**Comment by clarkdo** [comment]
If you want to serve static files on niginx, you can just move static to nginx and config the proxy.

```js
 location /staticx/ {
        root /www/static;
  }
```

Is there any concerns ?

**Comment by ishitatsuyuki** [comment]
Please read the issue content.

**Comment by cesasol** [comment]
Serving static files directly to nginx has nothing to do with the nuxt or node side, it must be done with the nginx config. There are multiple ways to achieve this, like selecting the types of files that you want to directly serve. Here is an example:

´´´
location ~* ^.+\.(jpg|svg|jpeg|gif|png|ico|css|zip|tgz|gz|rar|bz2|pdf|txt|tar|wav|bmp|rtf|js|flv|swf|html|htm)$ {
	gzip on;
	expires 365d;
	root   /absolute/path/to/sourcefiles/static;
}
´´´

**Comment by ishitatsuyuki** [comment]
Hey, please read the description thoroughly before commenting.

I'm talking about **documenting** which file are served (through `serve-static`) under `_nuxt/`. Undocumented thing may change and break any time, so documenting allows me to reliably move this job to nginx.

Also, it's not a simple path mount currently, but it also requires several filters so the server side source would not be revealed. Those needs to be either refactored to be moved out, or documented to be ignored.

**Comment by EmmanuelBeziat** [comment]
I strongly agree. Nuxt has a very strange behavior to create files in `.nuxt/dist` and to look for them in a non-existing `_nuxt` folder once the app is started. This is obviously odd. The two should match, and at least this behavior should be documented with an explanation on how to handle it.

**Comment by ishitatsuyuki** [comment]
Well, we have another option: copy the static files to dist/ just like SPA or generate mode do.

This is safer because we can change the internal behavior at any time, but a drawback is that it may be confusing to users (they may think it's completely static but it's not).

**Comment by EmmanuelBeziat** [comment]
Well, my goal would be to have Nothing more to do than a build and restart the app, with an automated process. Needing to manually select some files from different folders (static, .nuxt/dist, root...) is a big downside to me…

**Comment by yakovyarmo** [comment]
Did you manage to use ngnix to serve the static files?

**Comment by EmixMaxime** [comment]
Totally agree with @ishitatsuyuki 
The documentation is very confusing about serving static files...
I'll give it a look whether I could improve that part.

**Comment by ishitatsuyuki** [comment]
This is still a valid issue.

**Comment by LeoMartinDev** [comment]
Yes it is, it is not okay to serve static files using NodeJS...

**Comment by ishitatsuyuki** [comment]
To provide an update, we no longer need the exclusion because we now put them into a separate directory. It's much simpler, just need a documentation (and maybe changelogs) on which directory we should mount.

**Comment by ishitatsuyuki** [comment]
The bot is malfunctioning.

**Comment by kenberkeley** [comment]
What a pity to see most of the SSR websites on [Showcase | Awesome Nuxt.js](https://awesome-nuxt.js.org/resources/showcase.html) do not GZIP their JS / CSS, which results in enormous network loads. Some of them are running **serious** business, not for demo purposes.

The developers of these websites might have already configured `gzip on` on `nginx.conf` according to https://nuxtjs.org/faq/nginx-proxy . However, as you see, it's not working at all.

We tend to **trust** the official docs and we sincerely appreciate Nuxt team's hard work. However, Nuxt team should also test the functions before writing docs. Maybe it should be a caveat about GZIP on FAQ refers to https://github.com/nuxt/nuxt.js/issues/105 and this issue.

**Comment by manniL** [comment]
@kenberkeley I picked a random sample of 10 sites out of the showcases and **all** had either GZIP or Brotli encoding enabled. Please double-check 😉 


The deployment of a Nuxt app is up to the user. We can't set up NGINX/Apache for every use-case and can only provide ideas/configs that work in most of the cases.

Regarding the linked config itself, @pimlie can give you more info about that ☺️

**Comment by kenberkeley** [comment]
@manniL You should double-check. I picked ramdom sample of 10 sites and none of them had GZIP.

**Comment by kenberkeley** [comment]
All *Request Headers* include `accept-encoding: gzip, deflate, br` while all *Response Headers* do not include `content-encoding: gzip / br` .

**Comment by kenberkeley** [comment]
Don't trust me, but verify, please.

![image](https://user-images.githubusercontent.com/13287948/53330729-54cfb000-3943-11e9-9a71-82da61a4cf8f.png)

**Comment by manniL** [comment]
@kenberkeley must be on your side.

![image](https://user-images.githubusercontent.com/640208/53330965-b7935e00-38e7-11e9-9892-0eb36fcccda0.png)

**Comment by kenberkeley** [comment]
WT...

**Comment by kenberkeley** [comment]
Incognito mode, got the same result... Emmm... Maybe my Chrome broke...

![image](https://user-images.githubusercontent.com/13287948/53331495-24891100-3945-11e9-9a79-84e1a62d70d6.png)

**Comment by pimlie** [comment]
@kenberkeley The size of the file doesnt change due to gzip compression. Gzip compression is part of the http protocol to reduce transfer size, as soon as the file hits your browser the file is decompressed and saved in its normal format. So I think you are just looking at the wrong info, eg look at @manniL's screenshots and you will see that his size column also says `466KB` for the vendor bundle, but the transferred size is lower due to gzip compression.

**Comment by kenberkeley** [comment]
ESET Autivirus is the culprit! After uninstalling it:
![image](https://user-images.githubusercontent.com/13287948/53333024-d83fd000-3948-11e9-8bf8-d0be884e6574.png)

**Comment by altusgerona** [comment]
Is this issue rendered obsolete with `nuxt export`?

**Comment by manniL** [comment]
> 
> 
> Is this issue rendered obsolete with `nuxt export`?

Not necessarily, because it's mostly related to a deploy with a Node.js server (`nuxt build && nuxt start`)

**Comment by blacklizard** [comment]
if anyone is still having  this similar issue for SSR mode, you can use my nginx config. Nginx will serve all the static files.

In `nuxt.config.js` 

```js
build: {
  publicPath: '/assets/',
},
```

`nginx.v-host.conf`

```nginx
server {
  listen 80;
  server_name your_domain_name;
  root /path/to/your/project/.nuxt/dist/client;
  index index.html;

  gzip on;
  gzip_disable "msie6";
  gzip_vary on;
  gzip_proxied any;
  gzip_comp_level 6;
  gzip_buffers 16 8k;
  gzip_http_version 1.1;
  gzip_min_length 256;
  gzip_types
    application/atom+xml
    application/geo+json
    application/javascript
    application/x-javascript
    application/json
    application/ld+json
    application/manifest+json
    application/rdf+xml
    application/rss+xml
    application/xhtml+xml
    application/xml
    font/eot
    font/otf
    font/ttf
    image/svg+xml
    text/css
    text/javascript
    text/plain
    text/xml;

  location /assets {
    limit_except GET {
      deny all;
    }

    expires 365d;
    alias /path/to/your/project/.nuxt/dist/client;
  }

  location / {
    limit_except GET {
      deny all;
    }

    add_header Last-Modified $date_gmt;
    add_header Cache-Control 'no-store, no-cache, must-revalidate, proxy-revalidate, max-age=0';
    if_modified_since off;
    expires off;
    etag off;
    proxy_redirect                      off;
    proxy_set_header Host               $host;
    proxy_set_header X-Real-IP          $remote_addr;
    proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Proto  $scheme;
    proxy_read_timeout          1m;
    proxy_connect_timeout       1m;
    proxy_pass http://127.0.0.1:3000;
  }
}
```
